File: elo_rating.py
# For testing purposes only. Use utility.py package instead!
def get_new_rating(r0, opponents_rating, wins_first, wins_opponent, k=35):
    """

    :param r0: - старый рейтинг
    :param k:
    :return:
    """
    G = get_g(wins_first, wins_opponent)
    W = 0.5
    if wins_first > wins_opponent:
        W = 1
    elif wins_first < wins_opponent:
        W = 0
    We = get_we(r0, opponents_rating)
    return r0 + k * G * (W - We)


def get_g(wins_first, wins_second):
    difference = wins_first - wins_second
    if abs(difference) <= 1:
        return 1
    elif abs(difference) == 2:
        return 3 / 2
    else:
        return (11 + abs(difference)) / 8


def get_we(rating_first, rating_second):
    dr = rating_first - rating_second
    denominator = 10**(-dr / 400) + 1
    return 1 / denominator


if __name__ == "__main__":
    rating_A = 630
    rating_B = 500
    rating_C = 480
    A_wins = 3
    B_wins = 1
    print("team A vs team B")
    print("Situation first: A:3 - B:1")
    rating_A_new = get_new_rating(rating_A, rating_B, A_wins, B_wins)
    print(f"difference for A: {rating_A_new - rating_A}")
    rating_B_new = get_new_rating(rating_B, rating_A, B_wins, A_wins)
    print(f"difference for B: {rating_B_new - rating_B}")

    print("\nSituation second: A:1 - B:3")
    A_wins = 1
    B_wins = 3
    rating_A_new = get_new_rating(rating_A, rating_B, A_wins, B_wins)
    print(f"difference for A: {rating_A_new - rating_A}")
    rating_B_new = get_new_rating(rating_B, rating_A, B_wins, A_wins)
    print(f"difference for B: {rating_B_new - rating_B}")

    print("\nSituation third: A:2 - B:2")
    A_wins = 2
    B_wins = 2
    rating_A_new = get_new_rating(rating_A, rating_B, A_wins, B_wins)
    print(f"difference for A: {rating_A_new - rating_A}")
    rating_B_new = get_new_rating(rating_B, rating_A, B_wins, A_wins)
    print(f"difference for B: {rating_B_new - rating_B}")

    print("\n\nteam B vs team C")
    print("Situation first: B:3 - C:1")
    B_wins = 3
    C_wins = 1
    rating_B_new = get_new_rating(rating_B, rating_C, B_wins, C_wins)
    print(f"difference for B: {rating_B_new - rating_B}")
    rating_C_new = get_new_rating(rating_C, rating_B, C_wins, B_wins)
    print(f"difference for C: {rating_C_new - rating_C}")

    print("\nSituation second: B:1 - C:3")
    B_wins = 1
    C_wins = 3
    rating_B_new = get_new_rating(rating_B, rating_C, B_wins, C_wins)
    print(f"difference for B: {rating_B_new - rating_B}")
    rating_C_new = get_new_rating(rating_C, rating_B, C_wins, B_wins)
    print(f"difference for C: {rating_C_new - rating_C}")

    print("\nSituation third: B:2 - C:2")
    B_wins = 2
    C_wins = 2
    rating_B_new = get_new_rating(rating_B, rating_C, B_wins, C_wins)
    print(f"difference for B: {rating_B_new - rating_B}")
    rating_C_new = get_new_rating(rating_C, rating_B, C_wins, B_wins)
    print(f"difference for C: {rating_C_new - rating_C}")

File: elo_rating.go
package main

import (
	"fmt"
	"math"
)

func GetNewRating(r0 float64, opponentsRating float64, winsFirst int, winsOpponent int, k ...int) float64 {
	kCoef := 35
	if len(k) > 0 {
		kCoef = k[0]
	}
	var G float64 = getG(winsFirst, winsOpponent)

	var W float64 = 0.5
	if winsFirst > winsOpponent {
		W = 1.0
	} else if winsFirst < winsOpponent {
		W = 0
	}
	var We float64 = getWe(r0, opponentsRating)
	return r0 + float64(kCoef)*G*(W-We)
}

func getG(winsFirst int, winsSecond int) float64 {
	var difference float64 = float64(winsFirst) - float64(winsSecond)
	if math.Abs(difference) <= 1.0 {
		return 1.0
	} else if math.Abs(difference) == 2.0 {
		return 3.0 / 2.0
	}
	return (11.0 + math.Abs(difference)) / 8.0
}

func getWe(ratingsFirst float64, ratingsSecond float64) float64 {
	dr := ratingsFirst - ratingsSecond
	denominator := math.Pow(10, (-dr/400)) + 1.0
	return 1 / denominator
}

func main() {
	var ratingA float64 = 630
	var ratingB float64 = 500
	var ratingC float64 = 480
	fmt.Println("First situation: A (rating: 630) 3 vs 1 B (rating: 500)")
	AWins := 3
	BWins := 1
	ratingANew := GetNewRating(ratingA, ratingB, AWins, BWins)
	fmt.Println("difference for A: ", ratingANew-ratingA)
	ratingBNew := GetNewRating(ratingB, ratingA, BWins, AWins)
	fmt.Println("difference for B: ", ratingBNew-ratingB)

	fmt.Println("Second situation: A (rating: 630) 1 vs 3 B (rating: 500)")
	AWins = 1
	BWins = 3
	ratingANew = GetNewRating(ratingA, ratingB, AWins, BWins)
	fmt.Println("difference for A: ", ratingANew-ratingA)
	ratingBNew = GetNewRating(ratingB, ratingA, BWins, AWins)
	fmt.Println("difference for B: ", ratingBNew-ratingB)

	fmt.Println("Third situation: A (rating: 630) 2 vs 2 B (rating: 500)")
	AWins = 2
	BWins = 2
	ratingANew = GetNewRating(ratingA, ratingB, AWins, BWins)
	fmt.Println("difference for A: ", ratingANew-ratingA)
	ratingBNew = GetNewRating(ratingB, ratingA, BWins, AWins)
	fmt.Println("difference for B: ", ratingBNew-ratingB)

	fmt.Println("\n\nteam B vs team C")
	fmt.Println("First situation: B (rat. 500) 3 vs 1 C (rat.480)")
	BWins = 3
	CWins := 1
	ratingBNew = GetNewRating(ratingB, ratingC, BWins, CWins)
	fmt.Println("difference for B: ", ratingBNew-ratingB)
	ratingCNew := GetNewRating(ratingC, ratingB, CWins, BWins)
	fmt.Println("difference for C: ", ratingCNew-ratingC)

	fmt.Println("Second situation: B (rat. 500) 1 vs 3 C (rat.480)")
	BWins = 1
	CWins = 3
	ratingBNew = GetNewRating(ratingB, ratingC, BWins, CWins)
	fmt.Println("difference for B: ", ratingBNew-ratingB)
	ratingCNew = GetNewRating(ratingC, ratingB, CWins, BWins)
	fmt.Println("difference for C: ", ratingCNew-ratingC)

	fmt.Println("Third situation: B (rat. 500) 2 vs 2 C (rat.480)")
	BWins = 2
	CWins = 2
	ratingBNew = GetNewRating(ratingB, ratingC, BWins, CWins)
	fmt.Println("difference for B: ", ratingBNew-ratingB)
	ratingCNew = GetNewRating(ratingC, ratingB, CWins, BWins)
	fmt.Println("difference for C: ", ratingCNew-ratingC)
}
